base/compiler/ssair/passes.jl

```diff
@@ -921,15 +932,68 @@ function sroa_pass!(ir::IRCode)
         used_ssas = copy(compact.used_ssas)
         simple_dce!(compact, (x::SSAValue) -> used_ssas[x.id] -= 1)
         ir = complete(compact)
-        sroa_mutables!(ir, defuses, used_ssas, lazydomtree)
+        sroa_mutables!(ir, defuses, used_ssas, lazydomtree, inlining)
         return ir
     else
         simple_dce!(compact)
         return complete(compact)
     end
 end
 
-function sroa_mutables!(ir::IRCode, defuses::IdDict{Int, Tuple{SPCSet, SSADefUse}}, used_ssas::Vector{Int}, lazydomtree::LazyDomtree)
+function try_inline_finalizer!(ir::IRCode, argexprs::Vector{Any}, idx::Int, mi::MethodInstance, inlining::InliningState)
+    code = get(inlining.mi_cache, mi, nothing)
+    if code isa CodeInstance
+        if use_const_api(code)
+            # No code in the function - Nothing to do
+            inlining.et !== nothing && push!(inlining.et, mi)
+            return true
+        end
+        src = code.inferred
+    else
+        src = code
+    end
+
+    src = inlining_policy(inlining.interp, src, IR_FLAG_NULL, mi, Any[])
+    src === nothing && return false
+    src = retrieve_ir_for_inlining(mi, src)
+
+    # For now: Require finalizer to only have one basic block
+    length(src.cfg.blocks) == 1 || return false
+
+    # Ok, we're committed to inlining the finalizer
+    inlining.et !== nothing && push!(inlining.et, mi)
+
+    linetable_offset, extra_coverage_line = ir_inline_linetable!(ir.linetable, src, mi.def, ir[SSAValue(idx)][:line])
+    if extra_coverage_line != 0
+        insert_node!(ir, idx, NewInstruction(Expr(:code_coverage_effect), Nothing, extra_coverage_line))
+    end
+
+    # TODO: Use the actual inliner here rather than open coding this special
+    # purpose inliner.
```

Comment:
Seems like we should only emit the `Expr(:invoke, ...)` here (or create an `InliningTodo`)? Actually choosing to inline the code, vs do the lookup of the method, seem like separate decisions. And mixing it up here seems to be requiring an extra parameter here and other code duplication with inlining.

Reply:
In full generality yes, but that would require a late-inlining pass, for which our pass structure is not ready yet. Given the stringent restrictions already placed on this, I don't think it's a particularly big problem, but it would be cleaner of course to do the inlining in a separate pass.